src/main/java/net/fabricmc/loom/util/ZipUtils.java

```diff
@@ -0,0 +1,254 @@
+		}
+	}
+
+	public static void pack(Path from, Path zip) {
+		try {
+			Files.deleteIfExists(zip);
+		} catch (IOException e) {
+			throw new UncheckedIOException("Failed to add file to zip", e);
+		}
+
+		if (!Files.isDirectory(from)) throw new IllegalArgumentException(from + " is not a directory!");
+
+		try (StitchUtil.FileSystemDelegate fs = StitchUtil.getJarFileSystem(zip.toFile(), true)) {
+			for (Path fromPath : (Iterable<Path>) Files.walk(from)::iterator) {
+				if (!Files.isRegularFile(fromPath)) continue;
+				Path fsPath = fs.get().getPath(from.relativize(fromPath).toString());
+				Path fsPathParent = fsPath.getParent();
+				if (fsPathParent != null) Files.createDirectories(fsPathParent);
+				Files.copy(fromPath, fsPath, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.COPY_ATTRIBUTES);
+			}
+		} catch (IOException e) {
+			throw new UncheckedIOException("Failed to pack file to zip", e);
+		}
+	}
+
+	public static void add(Path zip, String path, byte[] bytes) {
+		add(zip, Collections.singleton(Pair.of(path, bytes)));
+	}
+
+	public static void add(Path zip, Iterable<Pair<String, byte[]>> files) {
+		try (StitchUtil.FileSystemDelegate fs = StitchUtil.getJarFileSystem(zip.toFile(), true)) {
+			for (Pair<String, byte[]> pair : files) {
+				Path fsPath = fs.get().getPath(pair.getLeft());
+				Path fsPathParent = fsPath.getParent();
+				if (fsPathParent != null) Files.createDirectories(fsPathParent);
+				Files.write(fsPath, pair.getRight(), StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
+			}
+		} catch (IOException e) {
+			throw new UncheckedIOException("Failed to add file to zip", e);
+		}
+	}
+
+	public static boolean replace(Path zip, String path, byte[] bytes) {
+		try {
+			replaceStrict(zip, path, bytes);
+			return true;
+		} catch (UncheckedIOException e) {
+			if (e.getCause() instanceof NoSuchFileException) {
+				return false;
+			}
+
+			throw e;
+		}
+	}
+
+	public static void replaceStrict(Path zip, String path, byte[] bytes) {
+		try (StitchUtil.FileSystemDelegate fs = StitchUtil.getJarFileSystem(zip.toFile(), true)) {
+			Path fsPath = fs.get().getPath(path);
+
+			if (Files.exists(fsPath)) {
+				Files.write(fsPath, bytes, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
+			} else {
+				throw new NoSuchFileException(fsPath.toString());
+			}
+		} catch (IOException e) {
+			throw new UncheckedIOException("Failed to replace file in zip", e);
+		}
+	}
+
+	public static boolean transformString(Path zip, Collection<Pair<String, UnsafeUnaryOperator<String>>> transforms) {
+		return transformString(zip, transforms.stream());
+	}
+
+	public static boolean transformString(Path zip, Stream<Pair<String, UnsafeUnaryOperator<String>>> transforms) {
+		return transformString(zip, transforms
+				.collect(Collectors.groupingBy(Pair::getLeft,
+						Collectors.mapping(Pair::getRight,
+								Collectors.reducing(null, MoreObjects::firstNonNull)))));
```

Comment:
Hmm, I wonder if this is a bit inefficient with the nested collectors...

Reply:
Shouldn't be too bad, if it is I can replace it with non-stream code